src/main/java/org/elasticsearch/common/util/concurrent/BaseFuture.java

```diff
@@ -118,6 +117,22 @@ public V get() throws InterruptedException, ExecutionException {
         return sync.get();
     }
 
+    /**
+     * Similar to {@link #get()} with the exception that it never blocks. If the operation wasn't completed yet
+     * it will return {@code null}. Otherwise it will behave in a similar fashion to {@link #get()} and will return
+     * the computed result, including throwing a {@link CancellationException} or {@link ExecutionException} if
+     * the computation was cancelled or threw an error (respectively).
+     *
+     * @throws CancellationException if the computation was cancelled
+     * @throws ExecutionException    if the computation threw an exception
+     */
+    public V tryGet() throws CancellationException, ExecutionException {
+        if (sync.isDone() == false) {
+            return null;
+        }
+        return sync.getValue();
+    }
```

Comment:
I had not though about option 2 but I really like the fact that it does not require to add a new API.

But if this argument does not resonate to you, feel free to push the change as-in, I don't have too strong feelings about it.